File: layers/src/main/java/com/psliusar/layers/ReplaceTransition.java
package com.psliusar.layers;

import java.util.ArrayList;

import androidx.annotation.NonNull;

public class ReplaceTransition<LAYER extends Layer<?>> extends Transition<LAYER> {

    private int lowestVisibleLayer = -1;
    private int replaceLayerIndex = -1;

    ReplaceTransition(@NonNull Layers layers, @NonNull Class<LAYER> layerClass) {
        super(layers, layerClass);
    }

    ReplaceTransition(@NonNull Layers layers, int index) {
        super(layers, index);
    }

    @Override
    protected void onTransition() {
        super.onTransition();

        // TODO replace layer by custom index

        // Add a new layer first
        stackEntry.inTransition = true;
        layers.commitStackEntry(stackEntry);

        final int stackSize = layers.getStackSize();
        lowestVisibleLayer = layers.getLowestVisibleLayer();
        replaceLayerIndex = stackSize - 2;
        if (replaceLayerIndex >= 0) {
            layers.getStackEntryAt(replaceLayerIndex).valid = false;
        }

        if (replaceLayerIndex >= lowestVisibleLayer) {
            // Mark layers for transition
            setTransitionState(lowestVisibleLayer);

            // Animate layers
            for (int i = lowestVisibleLayer; i < stackSize; i++) {
                animateLayer(layers.getStackEntryAt(i).layerInstance,
                    i == stackSize - 1 ? AnimationType.ANIMATION_UPPER_IN : AnimationType.ANIMATION_LOWER_OUT);
            }
        }
    }

    @Override
    protected void onAfterTransition() {
        super.onAfterTransition();

        if (replaceLayerIndex >= lowestVisibleLayer) {
            resetTransitionState(lowestVisibleLayer);
        }

        if (replaceLayerIndex >= 0) {
            layers.removeLayerAt(replaceLayerIndex);
        }
    }

    @Override
    protected void fastForward(@NonNull ArrayList<StackEntry> stack) {
        if (!isApplied()) {
            stack.add(stackEntry);
        }
        if (!isApplied() || hasAnimations()) {
            if (replaceLayerIndex >= 0) {
                stack.remove(replaceLayerIndex);
            }
        }
    }
}

File: layers/src/main/java/com/psliusar/layers/ReplaceTransition.kt
package com.psliusar.layers

import java.util.ArrayList

/**
 * Transition replaces [Layer] by the given index in the stack. Replaces Layer at the top of the
 * stack if the index is omitted.
 */
internal class ReplaceTransition<L : Layer>(
    layers: Layers,
    layerClass: Class<L>
) : Transition<L>(layers, StackEntry(layerClass)) {

    private var lowestVisibleLayer = -1
    private var replaceLayerIndex = -1

    override fun onTransition() {
        val insertedIndex = if (index == -1) {
            layers.stackSize
        } else {
            index + 1
        }
        replaceLayerIndex = insertedIndex - 1

        // Add a new layer first
        stackEntry.inTransition = true
        layers.addStackEntry(stackEntry, insertedIndex)

        lowestVisibleLayer = layers.getLowestVisibleLayer()
        if (replaceLayerIndex >= 0) {
            layers.getStackEntryAt(replaceLayerIndex).valid = false
        }

        // No animations for invisible changes
        if (lowestVisibleLayer < 0 || insertedIndex < lowestVisibleLayer) return

        // Mark layers for transition
        setTransitionState(lowestVisibleLayer)

        // Animate Layers
        val stackSize = layers.stackSize
        for (i in replaceLayerIndex until stackSize) {
            if (i < 0) continue

            val layer = layers.getStackEntryAt(i).layerInstance
                ?: throw IllegalStateException("Layer instance must exist")
            val anim = when {
                i == insertedIndex -> AnimationType.ANIMATION_UPPER_IN // New Layer
                i > insertedIndex -> continue // Do not animate Layers above the inserted one
                else -> AnimationType.ANIMATION_LOWER_OUT // Existing Layers below
            }
            animateLayer(layer, anim)
        }
    }

    override fun onAfterTransition() {
        if (replaceLayerIndex >= lowestVisibleLayer) {
            resetTransitionState(lowestVisibleLayer)
        }

        if (replaceLayerIndex >= 0) {
            layers.removeLayerAt<Layer>(replaceLayerIndex)
        }
    }

    override fun fastForward(stack: ArrayList<StackEntry>) {
        if (started) return

        if (index == -1) {
            stack.removeLastOrNull()
            stack.add(stackEntry)
        } else {
            stack.removeAt(index)
            stack.add(index, stackEntry)
        }
    }
}